packages/rum-recorder/src/domain/rrweb/mutationObserver.ts

```diff
@@ -26,11 +26,58 @@ import { MutationController } from './mutation'
 
 type WithSerializedTarget<T> = T & { target: NodeWithSerializedNode }
 
+const enum MutationObserverStatus {
+  WaitingForMutation,
+  MutationsPending,
+  Stopped,
+}
+type MutationObserverState =
+  | {
+      status: MutationObserverStatus.WaitingForMutation
+    }
+  | {
+      status: MutationObserverStatus.MutationsPending
+      pendingMutations: MutationRecord[]
+      cancelScheduledProcess: () => void
+    }
+  | {
+      status: MutationObserverStatus.Stopped
+    }
+
+/**
+ * Maximum duration to wait before processing mutations. If the browser is idle, mutations will be
+ * processed more quickly. If the browser is busy executing small tasks (ex: rendering frames), the
+ * mutations will wait MUTATION_PROCESS_MAX_DELAY milliseconds before being processed. If the
+ * browser is busy executing a longer task, mutations will be processed after this task.
+ */
+const MUTATION_PROCESS_MAX_DELAY = 100
+
 /**
  * Buffers and aggregate mutations generated by a MutationObserver into MutationPayload
  */
 export function startMutationObserver(controller: MutationController, mutationCallback: MutationCallBack) {
-  const observer = new MutationObserver(monitor((mutations) => processMutations(mutations, mutationCallback)))
+  let state: MutationObserverState = {
+    status: MutationObserverStatus.WaitingForMutation,
+  }
+
+  const observer = new MutationObserver(
+    monitor((mutations) => {
+      switch (state.status) {
+        case MutationObserverStatus.WaitingForMutation:
+          state = {
+            status: MutationObserverStatus.MutationsPending,
+            pendingMutations: mutations,
+            cancelScheduledProcess: scheduleMutationProcess(processPendingMutations),
+          }
+          break
+
+        case MutationObserverStatus.MutationsPending:
+          state.pendingMutations.push(...mutations)
+          break
+      }
+    })
+  )
```

Comment:
instead of adding extra complexity to the mutation observer, what about delegating the async processing to another component?
cf [experiment](https://github.com/DataDog/browser-sdk/commit/6bc702548b3b88cf564f1d4ea477e092cfd9c317)

Reply:
I'm fine with moving this logic to another module, but I would keep the "State" approach (in this new module), because it makes more sense to me. I'll see what I can do.